Step3 : 수강신청(DB 적용)

Mo-bile opened the pull request:
안녕하세요? 리뷰 요청드립니다!
DB적용 테이블를 고민하고, 이것을 jdbc템플릿을 이용해서 한땀한땀 만들고 테스트 코드 만드는데 생각보다 시간이 소요되었습니다
'ORM 프레임웍 활용할 떄의 설계, 개발 프로세스' 영상에서 하이버네이트의 ddl 생성을 보여주신 것 처럼 해야하나 할까 했다가 그냥 직접 한땀한땀 했습니다! 👍 

질문 드립니다
### 질문1. 음 관계형 DB 사고와 조금 거리를 두고 하기위해서는 의식적이 노력이 필요한 것일까요?
도메인 설계 후 DB 적용을 하려고 하는데 고민하고 생각할 때 가장 어려운 점은 관계형 DB 사고에 익숙해져서 저도 모르게 그런 쪽으로 생각이 흘러가는게 어려웠습니다.

처음에는 그냥 객체마다 모두 다 테이블을 별도로 두고 외래키 잡고 join하면 안되나? 이런생각도 하고 테이블 컬럼이 많이지면 괜히 마음이 불편해지곤 했었습니다

그러면서 공통의 규칙을 가지고 객체가 함께 생성되고 변화되고 삭제되어야하는 존재로 묶다가도, 수강신청이라는 행동을 중심으로 생각하기도 하는 등의 고민이 많았습니다.

음 관계형 DB 사고와 조금 거리를 두고 하기위해서는 의식적이 노력이 필요한 것일까요?

### 질문2. 금번 DB 적용관련해서 이런 방식의 구조가 괜찮은지 고민이 됩니다.
`entity` <-> `mapper` <-> `domain`

무언가 하면서도 계속해서 아 이거 과한거 같은데 생각들면서도
entity는 다른 테이블의 id를 가지지만 domain은 다른 객체 그 자체를 가지고 하는데 이것을 서로 어떻게 변환을 해야할지 생각하다가 mapper를 시도했습니다.
그런데 이용할수록 아 과한데?? 이런생각이 들었습니다.

그러다가 service layer가 entity 를 의존하는 문제가 생겨서 
`Service` 와 `Repository` 간에 `CourseDataAccess` 를 만들었는데요
이것도 하면서 아 이거 과한거같은데 생각이 지속적으로 들었습니다.

무언가 단순하게 하려고 했는데 복잡하게 접근한 것 같습니다.
어떤 방향으로 하면 좋을까요?

감사힙니다

Commit:
feat : 도메인 모델과 테이블 매핑하는 entity, mapper 개발 완료

Commit:
feat : mapper을 통해 object 와 테이블 매핑 기존 db접근 및 코드에 적용

Commit:
feat : entity 바탕으로 테이블 생성 및 테스트 데이터 메모

Commit:
refactor : 테이블 컬럼 컨벤션에 맞춰서 변경 반영

Commit:
feat : Session 테이블 및 Enrollment 테이블 CRUD 완료

Commit:
feat : EnrolledUser 테이블 관련 CURD 코드 완료

Commit:
fix : list mapper에 대한 로직 수정

Commit:
fix : 모킹 시 id 도 넣을 수 있게 수정

Commit:
feat : service layer에서 수강신청 기능 구현 및 테스트 완료

Commit:
docs : db 적용과정 메모

Commit:
refactor : 패키지 경로 수정

Commit:
feat : repository 와 service 사이에 layer를 추가해 손쉽게 데이터를 다룰 수있게 변경 후 CourseService의 enroll 메서드 수정

Commit:
docs : 개발 상황 메모

javajigi commented on src/main/java/nextstep/courses/infrastructure/mapper/SessionMapper.java:
```diff
@@ -0,0 +1,92 @@
+import nextstep.courses.infrastructure.entity.EnrollmentEntity;
+import nextstep.courses.infrastructure.entity.SessionEntity;
+
+public class SessionMapper {
```
entity <-> mapper <-> domain 방식으로 접근하면 지금과 같은 복잡도가 발생할 수 밖에 없음
수강생 중 같은 패턴을 적용한 [은정님의 코드](https://github.com/next-step/java-lms/pull/802/changes) 또한 비슷한 양상을 보임
단, 재영님 mapper와 비교해 보고 좀 더 단순하게 구현할 수는 없는지 고민해 보면 좋겠다.

javajigi replied:
매번 Mapper를 생성하는 것이 번거롭다면 https://mapstruct.org/ 와 같은 라이브러리 사용을 고려해 보는 것도 방법일 수 있겠다.

javajigi commented on src/main/java/nextstep/courses/infrastructure/mapper/SessionMapper.java:
```diff
@@ -0,0 +1,92 @@
+        return sessions;
+    }
+    
+    public static Session toModel(SessionEntity entity) throws CanNotCreateException {
```
SessionEntity를 Session으로 변환하는 부분은 SessionEntity가 담당하도록 구현하는 것도 한 가지 접근 방식일 수 있지 않을까?

Mo-bile replied:
mapper의 주요 메서드를 service에서 entity 구현 메서드 내부로 이동하여 service layer가 repository의 구현을 모르게끔 변경했습니다.

javajigi commented on src/main/java/nextstep/courses/infrastructure/repository/CourseDataAccess.java:
```diff
@@ -0,0 +1,78 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class CourseDataAccess {
```
너무 과한 접근으로 보여짐
이 객체를 제거하고 Service가 담당하도록 구현하는 것은 어떨까?

Mo-bile replied:
해당 부분을 제거하고 SessionService, EnrolledUserService 두 Service가 담당하게 변경하여 각각 repository에 대한 책임을 분배했습니다

javajigi commented on src/main/java/nextstep/courses/service/CourseService.java:
```diff
@@ -1,27 +1,30 @@
-        Course course = courseRepository.findById(courseId);
-        
-        Payment payment = new PaymentService().payment("id");
+    public void enroll(NsUser loginUser, long courseId, long sessionId) throws CanNotJoinException, CanNotCreateException {
```
이와 같이 구현할 경우 코스의 모든 강의(Session)을 매번 db에서 조회해야 한다.
성능상 좋은 접근 방법은 아니라 생각한다.
Course와 Session이 1:n 관계이고, Session의 courseId를 FK로 가지는 구조가 되기 때문에 수강신청하고자 하는 Session을 바로 생성해 수강 신청 여부를 판단하도록 구현해야 하지 않을까?
구현 위치 또한 CourseService가 아니라 SessionService를 추가하고 이곳에서 구현해야 하지 않을까?

Mo-bile replied:
네! 불필요한 조회가 되는 방식보다는 session에서 이용하는것이 바람직하다고 생각합니다.

javajigi commented on src/main/java/nextstep/courses/service/CourseService.java:
```diff
@@ -1,27 +1,30 @@
+        Payment payment = new PaymentService().payment("0");
         course.enrollCourse(new SessionApply(loginUser.getId(), payment), sessionId);
-        courseRepository.save(course);
+        courseDataAccess.updateEnrolledUsers(course, sessionId);
```
loginUser가 수강 신청을 하고 있는데 loginUser을 인자로 전달해 수강 신청이 가능하도록 구현해야하지 않을까?

Mo-bile replied:
수강신청 로직을 SessionService 로 떼내고 loginUser 를 전달하게 했습니다!

javajigi reviewed (changes requested):
entity <-> mapper <-> domain 접근 방식으로 구현했군요. 👍 
이 접근 방식이 domain 객체를 훼손하지 않는 좋은 접근 방식이기는 하지만 중복과 복잡도가 증가한다는 단점이 있어요.
피드백 남겼으니 한번 검토해 보면 좋겠습니다.
추가로 db 매핑을 하면서 수강 신청 기능을 기존과 같이 Course로부터 접근했을 때 성능상 이슈가 발생할 것 같아요.
다른 방식으로 접근해보면 좋겠고요.
수강 신청 로직 구현에 잘못 구현하고 있는 부분이 있는 것 같아 피드백 남겼어요.

Q1. 음 관계형 DB 사고와 조금 거리를 두고 하기위해서는 의식적이 노력이 필요한 것일까요?
A1. 테이블 설계는 관계형 사고, 도메인 객체 설계는 oop 사고 역량을 쌓아야 하고요.
이 각각의 역량을 쌓은 후 의식적인 연습이 필요한 부분은 관계형 테이블과 OOP로 설계한 객체간의 불일치를 인식하고, 이 불일치하는 부분을 어떻게 해소할 것인지에 대한 의식적인 연습이 필요한 것 같아요.

질문2는 피드백에 남겼어요.

Commit:
refactor
1. mapper 내 메서드 명 변경
2. CourseDataAccess 제거하고 SessionService 및 CourseService 추가

Commit:
refactor : repository의 메서드 시그니처와 리턴타입 변경 후 mapper를 구현하는 메서드 내부로 이동하여 service layer에서 mapper 역할을 제거

Commit:
fix : mapper 예외처리 추가

Commit:
refactor : 수강신청을 SessionService로 완전히 이관

Commit:
refactor : 적절한 변수명으로 변경

Commit:
docs : 리팩터링 과정 메모

Mo-bile commented:
안녕하세요? 리뷰 요청드립니다. 👍 

금번 step 의 과제를 할수록 개발한 도메인에 db 적용이 꽤나 쉽지 않다는 것을 깨달았습니다.
RDB와 OOP 방식으로 로 설계한 객체 간에 불일치 하는 부분을 해소하는 "코드 구현 부분"에서 생각보다 많은 시간을 사용한 것 같습니다.

구현을 할수록 과제 요구사항인 `CRUD 쿼리와 코드를 구현하는데 집중하기 보다 테이블을 설계하고 객체 매핑하는 부분에 집중한다.` 이 요구사항과 별개의 내용에 집중하면서 시간을 사용하는 것 같아서 다소 아쉬웠습니다. 왜 그런지 고민해보니 처음부터 구조를 복잡하게 짜서 그런가싶기도 하고, 실무에서는 라이브러리 위주로 사용하다가 직접 구현을 하려고하니 녹록치 않았던 것 같기도 했습니다.

✓ 추가
아래 내용(join 도입 전 고민내용) 같이 join 을 할지 관련 내용을 남긴 뒤, join으로 조회한 방식으로 변경 후 커밋했습니다. 
이렇게 하니 repository 간 결합도를 낮추고, n+1 문제를 해결할 수있게 되었습니다.

```
// join 도입 전 고민내용
`enrollmentRepository`의 `findBySessionId()` 내부에서 `enrolledUserRepository`의 메소드를 곧 바로 호출하여 repository간 결합도가 높아졌는데요... 바로 join을 하면되나? 이런 고민도 되었습니다
그 이유가 
- repository 내부에 mapper 를 두고 반환하기
- domain이나 service 레이어에 repository 관련 관심사인 entity 노출안시키기
이런 부분에 집중하다 보니 고민이 되었던 것 같습니다...
```

Commit:
refactor : 기존 db 를 두번 조회하여 repository간 결합된 형태를 join으로 변경하고 n+1 문제 방지

Commit:
refactor : mapper 내 사용하지않는 불필요한 메서드 제거

javajigi commented on src/main/java/nextstep/courses/service/EnrolledUserService.java:
```diff
@@ -0,0 +1,49 @@
+        this.enrolledUserRepository = enrolledUserRepository;
+    }
+
+    public void updateEnrolledUsers(Session session) {
```
매번 수강 신청은 1명이 진행하는데 너무 복잡하게 접근하는 것은 아닐까?
수강 신청하는 강의(session id)와 수강 신청하는 사용자(user id) 값만 받아 enrolledUserRepository을 이용해 저장하면 끝이지 않을까?
매번 수강 신청할 때마다 전체 수강생 목록을 조회해 아래와 같이 구현하는 것은 성능상 이슈도 발생할 가능성이 높아 보임

Mo-bile replied:
@javajigi 
하나 질문드려도 될까요?
`session.enrollSession()`의 비즈니스 로직에서 유저가 수강신청을 성공하면 EnrolledUsers 객체의 enrolledUserList 필드에 유저가 추가 됩니다. 이렇게 추가된 유저가 누구인지 찾은 후 DB 에 insert 하는 JPA의 `dirty checking` 방식으로 접근했는데, 지금 방식에서는 너무 복잡하게 접근한 것일까요?

말씀주신 의도는 단순하게 접근하여 수강신청 성공 시 enrolledUserList 객체 필드에 유저가 추가 되더라도, 별도로 DB 에 insert 할 때 그 유저의 id를 인자로 넣어 Insert 하는게 어떻냐는 방식으로 이해가 됩니다. 이 방식으로 말씀주신게 맞으실까요?

당초 제 방식은 객체에 수강신청 추가 된 user가 누구인지 찾은 뒤 insert 하는 방식이 `OOP 스럽고, DB 적용과 자연스럽다`고 생각해서 당초의 방식으로 접근했었습니다 (물론 제가 구현한 방식은 복잡성과 성능상 이슈는 트레이드 오프가 수반된다고 생각합니다.) 👍

javajigi replied:
@Mo-bile 
JPA의 dirty checking 방식을 직접 구현했는지는 생각도 못했네요.
학습적인 측면으로는 긍정적이라 생각되네요.
JPA를 쓰지 않는 상태에서 지금과 같이 구현하다보니 복잡도가 너무 높아보여 피드백 남겼는데요.
피드백을 반영할 것인지 여부는 재영님이 결정하면 될 것 같아요.

javajigi commented on src/main/java/nextstep/courses/infrastructure/mapper/SessionMapper.java:
```diff
@@ -0,0 +1,110 @@
+import nextstep.courses.infrastructure.entity.EnrollmentEntity;
+import nextstep.courses.infrastructure.entity.SessionEntity;
+
+public final class SessionMapper {
```
역시나 SessionMapper의 복잡도는 높음
도메인 객체와 Entity 사이에 Mapper를 추가하는 방식으로 구현하면 중복도 많아지고 복잡도도 높아짐
Session과 같이 객체의 복잡도가 높은 부분만 Entity를 추가하고 Course, EnrolledUser와 같이 필드 수도 적고 복잡도가 낮은 객체는 도메인 객체를 직접 db 매핑에 활용하는 것은 어떨까?

Mo-bile replied:
제안 감사드립니다
저는 entity 가 있으면 mapper 가 반드시 필요하다고 생각했고, 이와 더불어 db 이용시에는 도메인을 이용하면 안된다고 생각했습니다.

그런데 고민해보니 Repository layer의 각 메서드 내에서 어차피 mapper 로 entity <-> doamin 변환하는데, 굳이 이 절차없이 곧바로 객체를 이용해서 조회를 해도 괜찮구나 깨닫게 되었습니다 감사합니다.

javajigi reviewed (approved):
Entity <-> Mapper <-> Domain  패턴이 도메인 객체에 영향을 최소한으로 미치지만 역시나 복잡도는 높아지는 것 같아요.
그럼에도 불구하고 미션 진행 👍 
피드백에도 남겼는데요. 도메인 객체의 필드 수가 많고, Object Graph의 복잡도가 높은 객체만 Entity를 추가하는 방식으로 접근하면 어떨까요?
모든 곳을 이 방식으로 구현하기에는 구현 비용이 너무 크네요. ㅠㅠㅠ
수강 신청 로직과 관련해 너무 복잡하게 접근하는 것 같아 피드백 남겼으니 4단계 진행할 때 함께 반영해 보세요.
